core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -62,14 +62,46 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
   def fromSeq[A](as: Seq[A]): Option[NonEmptyLazyList[A]] =
     if (as.nonEmpty) Option(create(LazyList.from(as))) else None
 
-  def fromLazyListPrepend[A](a: A, ca: LazyList[A]): NonEmptyLazyList[A] =
-    create(a +: ca)
+  def fromLazyListPrepend[A](a: => A, ll: => LazyList[A]): NonEmptyLazyList[A] =
+    create(a #:: ll)
 
-  def fromLazyListAppend[A](ca: LazyList[A], a: A): NonEmptyLazyList[A] =
-    create(ca :+ a)
+  def fromLazyListAppend[A](ll: => LazyList[A], a: => A): NonEmptyLazyList[A] =
+    create(ll #::: a #:: LazyList.empty)
 
   def apply[A](a: => A, as: A*): NonEmptyLazyList[A] =
-    create(LazyList.concat(LazyList(a), LazyList.from(as)))
+    create(a #:: LazyList.from(as))
+
+  // allows the creation of fully lazy `NonEmptyLazyList`s by prepending to this
+  def maybe[A](ll: => LazyList[A]): Maybe[A] = new Maybe(() => ll)
+
+  final class Maybe[A] private[NonEmptyLazyList] (private[this] var mkLL: () => LazyList[A]) {
+    // because instances of this class are created explicitly, they might be
+    // reused, and we don't want to re-evaluate `mkLL`
+    private[this] lazy val ll = {
+      val res = mkLL()
+      mkLL = null // allow GC
+      res
+    }
```

Comment:
Just to clarify (perhaps I'm missing something).
Wouldn't it be the same thing if we simply make `mkLL` a `lazy val` itself?
```suggestion
  final class Maybe[A] private[NonEmptyLazyList] (private[this] lazy val mkLL: () => LazyList[A]) {
```

Reply:
we want a lazy `LazyList[A]`, not a lazy `() => LazyList[A]`. also, can you have lazy ctor parameters? that doesn't seem like it would work to me, as they have to be evaluated to be passed to the ctor

Reply:
Ah, I see what you mean, thanks for the clarification.
Yes, I forgot that Scala does not allow `lazy val` for ctor params.

However, looks like your code tries to do an optimization that the compiler can handle by itself.
Consider this example:
```scala
object TestApp extends App {
  final class Foo(a0: () => String) {
    private lazy val a = a0()

    def printMe(): Unit = println(s">>> a = $a")
   }
}
```
Now check out how compiler handles it:
```java
   public static final class Foo {
        private String a;
        private Function0<String> a0;
        private volatile boolean bitmap$0;

        private String a$lzycompute() {
            Foo foo = this;
            synchronized (foo) {
                if (!this.bitmap$0) {
                    this.a = (String)this.a0.apply();
                    this.bitmap$0 = true;
                }
            }
            this.a0 = null;
            return this.a;
        }

        private String a() {
            if (!this.bitmap$0) {
                return this.a$lzycompute();
            }
            return this.a;
        }

        public void printMe() {
            Predef$.MODULE$.println((Object)(">>> a = " + this.a()));
        }

        public Foo(Function0<String> a0) {
            this.a0 = a0;
        }
    }
```
Looks like it does the same thing as the code you've written here (i.e. sets `a0` to null after initializing the `lazy val a` because it is no longer used anywhere in the code).

I checked it with Scala 2.13.11, 2.12.18 and 3.3.0. The latter produces a way more sophisticated code for the initialization procedure, but still sets `a0` to `null` eventually.

Reply:
really? huh. I wonder if that wasn't the case before 2.13.0, but things changed? because the thing I did was just taken from `LazyList`

Reply:
actually, does it even need to be a `() => LazyList[A]` anymore then? can it just be by name?